src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModule.java

```diff
@@ -127,7 +133,32 @@ public APlusProject getProject() {
     return project.getModuleManager().findModuleByName(getName());
   }
 
+  @Override
+  public boolean hasLocalChanges() {
+    VirtualFile virtualFile = VfsUtil.findFile(getFullPath(), true);
+    final boolean[] hasChanges = {false};
+    long downloadedAt = getDownloadedAt().toInstant().toEpochMilli();
+
+    if (virtualFile != null) {
+      VfsUtilCore.visitChildrenRecursively(virtualFile, new VirtualFileVisitor<Object>() {
```

Comment:
do you think this should be deanonymized and tested on the lower lever (`visitFile(...)`)?

Reply:
A good thing about that approach would be that `hasChanges` could be then a (`volatile`?) field of the de-anonymized (that single-cell array is a bit hacky, IMO, and also data race prone if `visitChildrenRecursively` use parallelism now or in future).
 
Another thing: I appreciate the effort you have put to it to make the traversal fast-quitting if there's a single modified file. I'd, however, like if more of the logic of this method could be pulled up to `Module` class. What would you like about such a design where `Module` class implemented `hasLocalChanges` but called subclass method `getLastModifiedTimestamp` that would simply calculate the maximum of file timestamps? Taking into account how seldom this method is called, I'd consider it okay to sacrifice fast-quit for clearer design.

Reply:
I would not call this hack a "fast-quit" because of [this](https://github.com/Aalto-LeTech/intellij-plugin/pull/195#discussion_r433709296), so it's more of a "do not traverse" option. 

I've now extracted the anonymous class into a regular one, would it be OK if I leave this part "as is" now? You are naturally free to tweak it into a better design (or otherwise I'd need some more explanations on how to do that myself) :pray:

Reply:
You're right, my assumption on how that class works was misguided.

And yes, this is very good now. Thank you for your refactoring effort. ❤️ I appreciate that much!